bash/bash/doc/bash.1
.\"
.\" MAN PAGE COMMENTS to
.\"
.\"	Chet Ramey
.\"	Information Network Services
.\"	Case Western Reserve University
.\"	chet@ins.CWRU.Edu
.\"
.\"	Last Change: Mon Jul 15 15:20:56 EDT 2002
.\"
.\" bash_builtins, strip all but Built-Ins section
.if \n(zZ=1 .ig zZ
.if \n(zY=1 .ig zY
.TH BASH 1 "2002 July 15" "GNU Bash-2.05b"
.\"
.\" There's some problem with having a `@'
.\" in a tagged paragraph with the BSD man macros.
.\" It has to do with `@' appearing in the }1 macro.
.\" This is a problem on 4.3 BSD and Ultrix, but Sun
.\" appears to have fixed it.
.\" If you're seeing the characters
.\" `@u-3p' appearing before the lines reading
.\" `possible-hostname-completions
.\" and `complete-hostname' down in READLINE,
.\" then uncomment this redefinition.
.\"
.de }1
.ds ]X \&\\*(]B\\
.nr )E 0
.if !"\\$1"" .nr )I \\$1n
.}f
.ll \\n(LLu
.in \\n()Ru+\\n(INu+\\n()Iu
.ti \\n(INu
.ie !\\n()Iu+\\n()Ru-\w\\*(]Xu-3p \{\\*(]X
.br\}
.el \\*(]X\h|\\n()Iu+\\n()Ru\c
.}f
..
.\"
.\" File Name macro.  This used to be `.PN', for Path Name,
.\" but Sun doesn't seem to like that very much.
.\"
.de FN
\fI\|\\$1\|\fP
..
.SH NAME
bash \- GNU Bourne-Again SHell
.SH SYNOPSIS
.B bash
[options]
[file]
.SH COPYRIGHT
.if n Bash is Copyright (C) 1989-2002 by the Free Software Foundation, Inc.
.if t Bash is Copyright \(co 1989-2002 by the Free Software Foundation, Inc.
.SH DESCRIPTION
.B Bash
is an \fBsh\fR-compatible command language interpreter that
executes commands read from the standard input or from a file.
.B Bash
also incorporates useful features from the \fIKorn\fP and \fIC\fP
shells (\fBksh\fP and \fBcsh\fP).
.PP
.B Bash
is intended to be a conformant implementation of the IEEE
POSIX Shell and Tools specification (IEEE Working Group 1003\.2).
.SH OPTIONS
In addition to the single-character shell options documented in the
description of the \fBset\fR builtin command, \fBbash\fR
interprets the following options when it is invoked:
.PP
.PD 0
.TP 10
.BI \-c "\| string\^"
If the
.B \-c
option is present, then commands are read from
.IR string .
If there are arguments after the
.IR string ,
they are assigned to the positional parameters, starting with
.BR $0 .
.TP
.B \-i
If the
.B \-i
option is present, the shell is
.IR interactive .
.TP
.B \-l
Make
.B bash
act as if it had been invoked as a login shell (see
.SM
.B INVOCATION
below).
.TP
.B \-r
If the
.B \-r
option is present, the shell becomes
.I restricted
(see
.SM
.B "RESTRICTED SHELL"
below).
.TP
.B \-s
If the
.B \-s
option is present, or if no arguments remain after option
processing, then commands are read from the standard input.
This option allows the positional parameters to be set
when invoking an interactive shell.
.TP
.B \-D
A list of all double-quoted strings preceded by \fB$\fP
is printed on the standard ouput.
These are the strings that
are subject to language translation when the current locale
is not \fBC\fP or \fBPOSIX\fP.
This implies the \fB\-n\fP option; no commands will be executed.
.TP
.B [\-+]O [\fIshopt_option\fP]
\fIshopt_option\fP is one of the shell options accepted by the
\fBshopt\fP builtin (see
.SM
.B SHELL BUILTIN COMMANDS
below).
If \fIshopt_option\fP is present, \fB\-O\fP sets the value of that option;
\fB+O\fP unsets it.
If \fIshopt_option\fP is not supplied, the names and values of the shell
options accepted by \fBshopt\fP are printed on the standard output.
If the invocation option is \fB+O\fP, the output is displayed in a format
that may be reused as input.
.TP
.B \-\-
A
.B \-\-
signals the end of options and disables further option processing.
Any arguments after the
.B \-\-
are treated as filenames and arguments.  An argument of
.B \-
is equivalent to \fB\-\-\fP.
.PD
.PP
.B Bash
also interprets a number of multi-character options.
These options must appear on the command line before the
single-character options to be recognized.
.PP
.PD 0
.TP
.B \-\-dump\-po\-strings
Equivalent to \fB\-D\fP, but the output is in the GNU \fIgettext\fP
\fBpo\fP (portable object) file format.
.TP
.B \-\-dump\-strings
Equivalent to \fB\-D\fP.
.TP
.B \-\-help
Display a usage message on standard output and exit successfully.
.TP
\fB\-\-init\-file\fP \fIfile\fP
.PD 0
.TP
\fB\-\-rcfile\fP \fIfile\fP
.PD
Execute commands from
.I file
instead of the standard personal initialization file
.I ~/.bashrc
if the shell is interactive (see
.SM
.B INVOCATION
below).
.TP
.B \-\-login
Equivalent to \fB\-l\fP.
.TP
.B \-\-noediting
Do not use the GNU
.B readline
library to read command lines when the shell is interactive.
.TP
.B \-\-noprofile
Do not read either the system-wide startup file
.FN /etc/profile
or any of the personal initialization files
.IR ~/.bash_profile ,
.IR ~/.bash_login ,
or
.IR ~/.profile .
By default,
.B bash
reads these files when it is invoked as a login shell (see
.SM
.B INVOCATION
below).
.TP
.B \-\-norc
Do not read and execute the personal initialization file
.I ~/.bashrc
if the shell is interactive.
This option is on by default if the shell is invoked as
.BR sh .
.TP
.B \-\-posix
Change the behavior of \fBbash\fP where the default operation differs
from the POSIX 1003.2 standard to match the standard (\fIposix mode\fP).
.TP
.B \-\-restricted
The shell becomes restricted (see
.SM
.B "RESTRICTED SHELL"
below).
.TP
.B \-\-verbose
Equivalent to  \fB\-v\fP.
.TP
.B \-\-version
Show version information for this instance of
.B bash
on the standard output and exit successfully.
.PD
.SH ARGUMENTS
If arguments remain after option processing, and neither the
.B \-c
nor the
.B \-s
option has been supplied, the first argument is assumed to
be the name of a file containing shell commands.
If
.B bash
is invoked in this fashion, 
.B $0
is set to the name of the file, and the positional parameters
are set to the remaining arguments.
.B Bash
reads and executes commands from this file, then exits.
\fBBash\fP's exit status is the exit status of the last command
executed in the script.
If no commands are executed, the exit status is 0.
An attempt is first made to open the file in the current directory, and,
if no file is found, then the shell searches the directories in
.SM
.B PATH
for the script.
.SH INVOCATION
A \fIlogin shell\fP is one whose first character of argument zero is a
.BR \- ,
or one started with the 
.B \-\-login
option.
.PP
An \fIinteractive\fP shell is one started without non-option arguments
and without the
.B \-c
option
whose standard input and error are
both connected to terminals (as determined by
.IR isatty (3)),
or one started with the
.B \-i
option.
.SM
.B PS1
is set and
.B $\-
includes
.B i
if
.B bash
is interactive,
allowing a shell script or a startup file to test this state.
.PP
The following paragraphs describe how
.B bash
executes its startup files.
If any of the files exist but cannot be read,
.B bash
reports an error.
Tildes are expanded in file names as described below under
.B "Tilde Expansion"
in the
.SM
.B EXPANSION
section.
.PP
When
.B bash
is invoked as an interactive login shell, or as a non-interactive shell
with the \fB\-\-login\fP option, it first reads and
executes commands from the file \fI/etc/profile\fP, if that
file exists.
After reading that file, it looks for \fI~/.bash_profile\fP,
\fI~/.bash_login\fP, and \fI~/.profile\fP, in that order, and reads
and executes commands from the first one that exists and is readable.
The
.B \-\-noprofile
option may be used when the shell is started to inhibit this behavior.
.PP
When a login shell exits,
.B bash
reads and executes commands from the file \fI~/.bash_logout\fP, if it
exists.
.PP
When an interactive shell that is not a login shell is started,
.B bash
reads and executes commands from \fI~/.bashrc\fP, if that file exists.
This may be inhibited by using the
.B \-\-norc
option.
The \fB\-\-rcfile\fP \fIfile\fP option will force
.B bash
to read and execute commands from \fIfile\fP instead of \fI~/.bashrc\fP.
.PP
When
.B bash
is started non-interactively, to run a shell script, for example, it
looks for the variable
.SM
.B BASH_ENV
in the environment, expands its value if it appears there, and uses the
expanded value as the name of a file to read and execute.
.B Bash
behaves as if the following command were executed:
.sp .5
.RS
.if t \f(CWif [ \-n "$BASH_ENV" ]; then . "$BASH_ENV"; fi\fP
.if n if [ \-n "$BASH_ENV" ]; then . "$BASH_ENV"; fi
.RE
.sp .5
but the value of the
.SM
.B PATH
variable is not used to search for the file name.
.PP
If
.B bash
is invoked with the name
.BR sh ,
it tries to mimic the startup behavior of historical versions of
.B sh
as closely as possible,
while conforming to the POSIX standard as well.
When invoked as an interactive login shell, or a non-interactive
shell with the \fB\-\-login\fP option, it first attempts to
read and execute commands from
.I /etc/profile
and
.IR ~/.profile ,
in that order.
The
.B \-\-noprofile
option may be used to inhibit this behavior.
When invoked as an interactive shell with the name
.BR sh ,
.B bash
looks for the variable
.SM
.BR ENV ,
expands its value if it is defined, and uses the
expanded value as the name of a file to read and execute.
Since a shell invoked as
.B sh
does not attempt to read and execute commands from any other startup
files, the
.B \-\-rcfile
option has no effect.
A non-interactive shell invoked with the name
.B sh
does not attempt to read any other startup files. 
When invoked as
.BR sh ,
.B bash
enters
.I posix
mode after the startup files are read.
.PP
When
.B bash
is started in
.I posix
mode, as with the
.B \-\-posix
command line option, it follows the POSIX standard for startup files.
In this mode, interactive shells expand the
.SM
.B ENV
variable and commands are read and executed from the file
whose name is the expanded value.
No other startup files are read.
.PP
.B Bash
attempts to determine when it is being run by the remote shell
daemon, usually \fIrshd\fP.
If
.B bash
determines it is being run by \fIrshd\fP, it reads and executes
commands from \fI~/.bashrc\fP, if that file exists and is readable.
It will not do this if invoked as \fBsh\fP.
The
.B \-\-norc
option may be used to inhibit this behavior, and the
.B \-\-rcfile
option may be used to force another file to be read, but
\fIrshd\fP does not generally invoke the shell with those options
or allow them to be specified.
.PP
If the shell is started with the effective user (group) id not equal to the
real user (group) id, and the \fB\-p\fP option is not supplied, no startup
files are read, shell functions are not inherited from the environment, the
.SM
.B SHELLOPTS
variable, if it appears in the environment, is ignored,
and the effective user id is set to the real user id.
If the \fB\-p\fP option is supplied at invocation, the startup behavior is

X11/ChangeLog

== 10.4 ChangeLog items ==

2005-03-09  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/darwin.c
	(DarwinUpdateModifiers): handle AlphaShift (caps lock) as a
	normal modifier, now that we disabled XKB (when merging with
	XFree86 4.4) that's the correct thing to do (4015381)

2005-03-01  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/create.c: more
	Xpm fixes, from
	https://bugs.freedesktop.org/show_bug.cgi?id=1920, attachment
	1909. (4031381)

	* xc/lib/Xpm/Imakefile: doh, forgot Imakefile portion of
	original patch

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/s_popen.c,
	xc/extras/Xpm/lib/parse.c, xc/extras/Xpm/lib/misc.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/WrFFrBuf.c,
	xc/extras/Xpm/lib/RdFToI.c, xc/extras/Xpm/lib/RdFToBuf.c,
	xc/extras/Xpm/lib/Imakefile, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/CrBufFrI.c, xc/extras/Xpm/lib/Attrib.c:
	merged changes from XFree86 HEAD branch (4022757)

2005-02-15  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz-server.c: request that
	deferred updates are disabled. Otherwise we lose 25%
	performance on the "cat file in xterm" test (3985081)

	Fix 3494176 (deadlock while closing OpenGL window):

	* xc/lib/GL/glx/glxext.c: make __glXLock a recursive mutex

	* xc/lib/GL/apple/dri_driver.h, xc/lib/GL/apple/dri_driver.c:
	make the screen mutex a pthread_mutex and make it recursive

2005-01-04  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/X11Controller.m: fixed several
	bugs in the customization dialog for the applications menu
	(3668769)

2004-11-16  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: added a signal
	handler to catch SIGTERM etc and exit cleanly, killing the X
	server and xinit (3879878)

2004-10-25  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/InfoPlist.strings:
	removed CFBundleShortVersionString and CFBundleGetInfoString
	(3851089)

2004-10-19  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/Localizable.strings:
	added "Terminal" to the strings file	

	* xc/programs/Xserver/hw/apple/X11Application.m (prefs_get):
	added code to call NSLocalizedString on each name in the
	default applications menu

	* xc/programs/Xserver/hw/apple/rootless-window.c
	(RootlessKnowsWindowNumber): added check for zero window id

2004-10-15  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/Localizable.strings:
	Expanded backslash escape sequences since .strings files don't
	seem to handle them correctly. (3833239)

2004-10-15  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/vfb/InitOutput.c: workaround 3834723
	by rearranging include order

2004-10-10  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/quartz.h,
	xc/programs/Xserver/hw/apple/quartz.c,
	xc/programs/Xserver/hw/apple/X11Controller.m: QuartzRunClient
	function now has a second arg - whether or not to quote the
	command (3816789, 3572554)

2004-10-08  John Harper  <jsh@apple.com>

	* xc/extras/Xpm/lib/scan.c, xc/extras/Xpm/lib/parse.c,
	xc/extras/Xpm/lib/hashtab.c, xc/extras/Xpm/lib/data.c,
	xc/extras/Xpm/lib/create.c, xc/extras/Xpm/lib/XpmI.h,
	xc/extras/Xpm/lib/WrFFrI.c, xc/extras/Xpm/lib/CrDatFrI.c,
	xc/extras/Xpm/lib/Attrib.c: merged Xpm changes from XFree86
	HEAD branch (3744542, 3778302)

2004-09-29  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/main.nib/: fixed
	typo (3820725)

2004-09-16  John Harper  <jsh@apple.com>

	* xc/programs/xterm/main.c: call openpty() to scan for a pseudo
	terminal (3768572)

	* xc/programs/xterm/xterm.man: fix documentation of -ls option
	(3803120)

2004-09-08  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/main.nib/: nib
	changes for 3313133 and 3316468

2004-09-03  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/English.lproj/main.nib/: gave
	applications dialog window a minimum size

	* xc/programs/Xserver/hw/apple/English.lproj/Localizable.strings:
	new file - ran genstrings on the .m files

	* xc/programs/Xserver/hw/apple/X11.pbproj/project.pbxproj:
	added Localizable.strings

	* xc/programs/Xserver/hw/apple/X11Controller.m,
	xc/programs/Xserver/hw/apple/X11Application.m: added comments
	for translators

	* xc/programs/Xserver/hw/apple/Imakefile: set OBJROOT and
	SYMROOT when calling xcodebuild

2004-08-25  John Harper  <jsh@apple.com>

	* quartz-wm/x-window.m: don't grab Mod1-grave for window
	cycling, only Mod2 (3774967)

2004-08-23  John Harper  <jsh@apple.com>

	* xc/programs/Xserver/hw/apple/bundle-main.c: fix typo in
	$XINITRC code (3621248)

2004-07-16  John Harper  <jsh@apple.com>

	* merged changes from XFree86 4.4.0 and fixed resulting
	brokenness (3570235, 3612446, 3548675, 3404805)

2004-06-10  John Harper  <jsh@apple.com>

	* xinitrc: quote shell expanded filenames to support home
	directories with spaces (3674992)

2004-03-29  John Harper  <jsh@apple.com>

	* FS/FSFontInfo.c, FS/FSFtNames.c, FS/FSGetCats.c,
	FS/FSListCats.c, FS/FSListExt.c, FS/FSOpenServ.c,
	FS/FSQGlyphs.c, FS/FSQXExt.c, FS/FSQXInfo.c, FS/FSlibint.h,
	FS/FSlibos.h, font/fc/fsconvert.c, font/fc/fserve.c,
	font/fc/fslibos.h, font/fontfile/dirfile.c,

ksh/ksh/src/cmd/ksh93/sh/bash.c
/*******************************************************************
*                                                                  *
*             This software is part of the ast package             *
*                Copyright (c) 1982-2004 AT&T Corp.                *
*        and it may only be used by you under license from         *
*                       AT&T Corp. ("AT&T")                        *
*         A copy of the Source Code Agreement is available         *
*                at the AT&T Internet web site URL                 *
*                                                                  *
*       http://www.research.att.com/sw/license/ast-open.html       *
*                                                                  *
*    If you have copied or used this software without agreeing     *
*        to the terms of the license you are infringing on         *
*           the license and copyright and are violating            *
*               AT&T's intellectual property rights.               *
*                                                                  *
*            Information and Software Systems Research             *
*                        AT&T Labs Research                        *
*                         Florham Park NJ                          *
*                                                                  *
*                David Korn <dgk@research.att.com>                 *
*                                                                  *
*******************************************************************/
/*
 * bash specific extensions
 * originally provided by Karsten Fleischer
 */

#include <string.h>

#include "defs.h"
#include "path.h"
#include "io.h"
#include "builtins.h"
#include "name.h"

#ifndef BASH_MAJOR
#   define BASH_MAJOR	"1"
#   define BASH_MINOR	"0"
#   define BASH_PATCH	"0"
#   define BASH_BUILD	"0"
#   define BASH_RELEASE	"experimental"
#endif 
#define BASH_VERSION	BASH_MAJOR "." BASH_MINOR "." BASH_PATCH "(" BASH_BUILD ")-" BASH_RELEASE 


void	sh_applyopts(Shopt_t);

extern const char	bash_pre_rc[];

static const char	e_bash_logout[] = "$HOME/.bash_logout";
static char *login_files[4];

const char sh_bash1[] =
	"[B?Enable brace group expansion. This option is only availabe in bash "
	"compatibility mode. In ksh mode, brace group expansion is always on.]"
	"[P?Do not follow symbolic links, use physical directory structure "
	"instead. Only available in bash compatibility mode.]";
const char sh_bash2[] =
"[l:login?Make the shell act as if it had been invoked as a login shell. "
"Only available if invoked as \bbash\b.]"
"[O]:?[shopt_option?\ashopt_option\a is one of the shell options accepted by "
	"the \bshopt\b builtin. If \ashopt_option\a is present, \b-O\b sets "
	"the value of that option; \b+O\b unsets it. If \ashopt_option\a is "
	"not supplied, the names and values of the shell options accepted by "
	"\bshopt\b are printed on the standard output. If the invocation "
	"option is \b+O\b, the output is displayed in a format that may be "
	"reused as input. Only available if invoked as \bbash\b.]"
"[01:init-file|rcfile]:[file?Execute commands from \afile\a instead of the "
	"standard personal initialization file ~/.bashrc if the shell is "
	"interactive. Only available if invoked as \bbash\b.]"
"[02:noediting?For option compatibility with \bbash\b only. Ignored.]"
"[03:noprofile?Do not read either the system-wide startup file or any of the "
	"personal initialization files. Only available if invoked as \bbash\b.]"
"[04:norc?Do not read and execute the personal initialization file "
	"~/.bashrc if the shell is interactive. Only available if invoked as "
	"\bbash\b.]"
"[05:posix?If invoked as \bbash\b, turn on POSIX compatibility. \bBash\b in "
	"POSIX mode is not the same as \bksh\b.]"
"[06:version?Print version number and exit.]";

const char sh_optshopt[] =
"+[-1c?\n@(#)$Id: shopt (AT&T Labs Research) 2003-02-13 $\n]"
"[-author?Karsten Fleischer <K.Fleischer@omnium.de>]"
USAGE_LICENSE
"[+NAME?shopt - set/unset variables controlling optional shell behavior]"
"[+DESCRIPTION?\bshopt\b sets or unsets variables controlling optional shell "
	"behavior. With no options, or with the \b-p\b option, a list of all "
	"settable options is displayed, with an indication of whether or not "
	"each is set.]"
"[p?Causes output to be displayed in a form that may be reused as input.]"
"[s?Set each \aoptname\a.]"
"[u?Unset each \aoptname\a.]"
"[q?Suppress output (quiet mode). The return status indicates whether the "
	"\aoptname\a is set or unset. If multiple \aoptname\a arguments are "
	"given with \b-q\b, the return status is zero if all \aoptname\as are "
	"enabled; non-zero otherwise.]"
"[o?Restricts the values of \aoptname\a to be those defined for the \b-o\b "
	"option to the set builtin.]"
"[+?If either \b-s\b or \b-u\b is used with no \aoptname\a arguments, the "
	"display is limited to those options which are set or unset.]"
"[+?\bshopt\b supports all bash options. Some settings do not have any effect "
	"or are are always on and cannot be changed."
"[+?The value of \aoptname\a must be one of the following:]{"
		"[+cdable_vars?If set, arguments to the \bcd\b command are "
			"assumed to be names of variables whose values are to "
			"be used if the usual \bcd\b proceeding fails.]"
		"[+cdspell?Currently ignored.]"
		"[+checkhash?Always on.]"
		"[+checkwinsize?Currently ignored.]"
		"[+cmdhist?Always on.]"
		"[+dotglob?If set, include filenames beginning with a \b.\b "
			"in the results of pathname expansion.]"
		"[+execfail?Always on.]"
		"[+expand_aliases?Always on.]"
		"[+extglob?Enable extended pattern matching features.]"
		"[+histappend?Always on.]"
		"[+histreedit?If set and an edit mode is selected, the user "
			"is given the opportunity to re-edit a failed history "
			"substitution.]"
		"[+histverify?If set and an edit mode is selected, the result "
			"of a history substitution will not be executed "
			"immediately but be placed in the edit buffer for "
			"further modifications.]"
		"[+hostcomplete?Currently ignored.]"
		"[+huponexit?Currently ignored.]"
		"[+interactive_comments?Always on.]"
		"[+lithist?Always on.]"
		"[+login_shell?This option is set if the shell is started as "
			"a login shell. The value cannot be changed.]"
		"[+mailwarn?Currently ignored.]"
		"[+no_empty_cmd_completion?Always on.]"
		"[+nocaseglob?Match filenames in a case-insensitive fashion "
			"when performing filename expansion.]"
		"[+nullglob?Allows filename patterns which match no files to "
			"expand to a null string, rather than themselves.]"
		"[+progcomp?Currently ignored.]"
		"[+promptvars?Currently ignored.]"
		"[+restricted_shell?This option is set if the shell is started "
			"as a restricted shell. The value cannot be changed. "
			"It is not reset during execution of startup files, "
			"allowing the startup files to determine whether the "
			"shell is restricted.]"
		"[+shift_verbose?Currently ignored.]"
		"[+sourcepath?If set, the \b.\b builtin uses the value of PATH "
			"to find the directory containing the file supplied "
			"as an argument.]"
		"[+xpg_echo?If set, the \becho\b and \bprint\b builtins "
			"expand backslash-escape sequences.]"
"}"
"\n"
"\n[optname ...]\n"
"\n"
"[+EXIT STATUS?]{"
	"[+?The return status when listing options is zero if all \aoptnames\a "
	"are enabled, non-zero otherwise. When setting or unsetting options, "
	"the return status is zero unless an \aoptname\a is not a valid shell "
	"option.]"
"}"

"[+SEE ALSO?\bset\b(1)]"
;

/* GLOBIGNORE discipline. Turn on SH_DOTGLOB on set, turn off on unset. */

static void put_globignore(register Namval_t* np, const char *val, int flags, Namfun_t *fp)
{
	if(val)
		sh_onoption(SH_DOTGLOB);
	else
		sh_offoption(SH_DOTGLOB);

	nv_putv(np,val,flags,fp);
}

const Namdisc_t SH_GLOBIGNORE_disc  = { sizeof(Namfun_t), put_globignore };

/* FUNCNAME discipline */

struct	funcname
{
	Namfun_t	hdr;
};

static void put_funcname(register Namval_t* np,const char *val,int flags,Namfun_t *fp)
{
	/* bash silently returns with an error when FUNCNAME is set,
	   unsetting FUNCNAME is allowed */
	if(val && !(flags&NV_RDONLY))
		error_info.exit(1);

	nv_putv(np,val,flags,fp);
}

const Namdisc_t SH_FUNCNAME_disc  = { sizeof(struct funcname), put_funcname };

#define	SET_SET		1
#define	SET_UNSET	2
#define	SET_NOARGS	4

/* shopt builtin */

int     b_shopt(int argc,register char *argv[],void *extra)
{
        Shell_t *shp = (Shell_t*)extra;
	int n, ret=0;
	Shopt_t newflags=shp->options, opt;
	int verbose=PRINT_SHOPT|PRINT_ALL|PRINT_NO_HEADER|PRINT_VERBOSE;
	int setflag=0, quietflag=0, oflag=0;
	memset(&opt,0,sizeof(opt));
#ifdef SHOPT_RAWONLY
	on_option(&newflags,SH_VIRAW);
#endif
	while((n = optget(argv,sh_optshopt)))
	{
		switch(n)
		{
		case 'p':
			verbose&=~PRINT_VERBOSE;
			break;
		case 's':
		case 'u':
			setflag|=n=='s'?SET_SET:SET_UNSET;
			if(setflag==(SET_SET|SET_UNSET))
			{
				errormsg(SH_DICT,ERROR_ERROR,"cannot set and unset options simultaneously");
				error_info.errors++;
			}
			break;
		case 'q':
			quietflag=1;
			break;
		case 'o':
			oflag=1;
			verbose&=~PRINT_SHOPT;
			break;
		case ':':
			errormsg(SH_DICT,2, "%s", opt_info.arg);
			continue;
		case '?':
			errormsg(SH_DICT,ERROR_usage(0), "%s", opt_info.arg);
			return(-1);
		}
	}
	if(error_info.errors)
		errormsg(SH_DICT,ERROR_usage(2),"%s",optusage(NIL(char*)));
	argc -= opt_info.index;
	if(argc==0)
	{
		/* no args, -s => mask=current options, -u mask=~(current options)
		   else mask=all bits */
		if(setflag&SET_SET)
			opt=newflags;
		else if(setflag&SET_UNSET)
			for(n=0;n<4;n++)
				opt.v[n]=~newflags.v[n];
		else
			memset(&opt,0xff,sizeof(opt));
		setflag=SET_NOARGS;
	}
	while(argc>0)
	{
		n=sh_lookup(argv[opt_info.index],shtab_options);
		if(n<=0||(setflag 
			&& (is_option(&opt,SH_INTERACTIVE)
			    || is_option(&opt,SH_RESTRICTED)
			    || is_option(&opt,SH_RESTRICTED2)
			    || is_option(&opt,SH_BASH)
			    || is_option(&opt,SH_LOGIN_SHELL)))
			||(oflag&&(n&SH_BASHOPT)))
		{
			errormsg(SH_DICT,ERROR_ERROR, e_option, argv[opt_info.index]);
			error_info.errors++;
			ret=1;
		}
		else
			on_option(&opt,n&0xff);
		opt_info.index++;
		argc--;
	}
	if(setflag&(SET_SET|SET_UNSET))
	{
		if(setflag&SET_SET)
		{
			if(sh_isoption(SH_INTERACTIVE))
				off_option(&opt,SH_NOEXEC);
			if(is_option(&opt,SH_VI)||is_option(&opt,SH_EMACS)||is_option(&opt,SH_GMACS))
			{
				off_option(&newflags,SH_VI);
				off_option(&newflags,SH_EMACS);
				off_option(&newflags,SH_GMACS);
			}
			for(n=0;n<4;n++)
				newflags.v[n] |= opt.v[n];
		}
		else if(setflag&SET_UNSET)
			for(n=0;n<4;n++)
				newflags.v[n] &= ~opt.v[n];
		sh_applyopts(newflags);
		shp->options = newflags;
		if(is_option(&newflags,SH_XTRACE))
			sh_trace(argv,1);
	}
	else if(!(setflag&SET_NOARGS)) /* no -s,-u but args, ret=0 if opt&mask==mask */
	{
		for(n=0;n<4;n++)
			ret+=((newflags.v[n]&opt.v[n])!=opt.v[n]);
	}
	if(!quietflag&&!(setflag&(SET_SET|SET_UNSET)))
		sh_printopts(newflags,verbose,&opt);
	return(ret);
}

/* mode = 0: init, called two times
        before parsing shell args with SH_PREINIT state turned on
	second time after sh_init() is through and with SH_PREINIT state turned off
   mode > 1: re-init
   mode < 0: shutdown
*/

void bash_init(int mode)
{
	Sfio_t		*iop;
	Namval_t	*np;
	int		n=0,xtrace,verbose;
	if(mode>0)
		goto reinit;
	if(mode < 0)
	{
		/* termination code */
		int fdin;
		if(sh_isoption(SH_LOGIN_SHELL) && !sh_isoption(SH_POSIX)
#ifdef PATH_BFPATH
			&& (fdin = path_open(sh_mactry((char*)e_bash_logout),NIL(Pathcomp_t*))) >= 0)
#else
			&& (fdin = path_open(sh_mactry((char*)e_bash_logout),"")) >= 0))
#endif		
		{
			char buff[IOBSIZE+1];
			sh_offstate(SH_INTERACTIVE);
			sh_onstate(SH_PROFILE);
			error_info.id = path_basename(e_bash_logout);
			iop = sfnew(NIL(Sfio_t*),buff,IOBSIZE,fdin,SF_READ);
			sh_eval(iop,0);
		}
		return;	
	}

	if(sh_isstate(SH_PREINIT))
	{	/* pre-init stage */
		if(sh_isoption(SH_RESTRICTED))
			sh_onoption(SH_RESTRICTED2);
		sh_onoption(SH_HISTORY2);
		sh_onoption(SH_INTERACTIVE_COMM);
		sh_onoption(SH_SOURCEPATH);
		sh_onoption(SH_HISTAPPEND);
		sh_onoption(SH_CMDHIST);
		sh_onoption(SH_LITHIST);
		sh_onoption(SH_NOEMPTYCMDCOMPL);
		if(sh.login_sh==2)
			sh_onoption(SH_LOGIN_SHELL);
		if(strcmp(astgetconf("CONFORMANCE",0,0,0),"standard")==0)
			sh_onoption(SH_POSIX);
		if(strcmp(astgetconf("UNIVERSE",0,0,0),"att")==0)
			sh_onoption(SH_XPG_ECHO);
		else
			sh_offoption(SH_XPG_ECHO);
		if(strcmp(astgetconf("PATH_RESOLVE",0,0,0),"physical")==0)
			sh_onoption(SH_PHYSICAL);
		else
			sh_offoption(SH_PHYSICAL);

		/* add builtins */
		sh_addbuiltin("shopt", b_shopt, &sh);

		/* set up some variables needed for --version
		 * needs to go here because --version option is parsed before the init script.
		 */
		if(np=nv_open("HOSTTYPE",sh.var_tree,0))
			nv_putval(np, BASH_HOSTTYPE, NV_NOFREE);
		if(np=nv_open("MACHTYPE",sh.var_tree,0))
			nv_putval(np, BASH_MACHTYPE, NV_NOFREE);
		if(np=nv_open("BASH_VERSION",sh.var_tree,0))
			nv_putval(np, BASH_VERSION, NV_NOFREE);
		if(np=nv_open("BASH_VERSINFO",sh.var_tree,0))
		{
			char *argv[7];
			argv[0] = BASH_MAJOR;
			argv[1] = BASH_MINOR;
			argv[2] = BASH_PATCH;
			argv[3] = BASH_BUILD;
			argv[4] = BASH_RELEASE;
			argv[5] = BASH_MACHTYPE;
			argv[6] = 0;
			nv_setvec(np, 0, 6, argv);
			nv_onattr(np,NV_RDONLY);
		}
		return;
	}

	/* rest of init stage */

	/* restrict BASH_ENV */
	if(np=nv_open("BASH_ENV",sh.var_tree,0))
	{
		const Namdisc_t *dp = nv_discfun(NV_DCRESTRICT);
		Namfun_t *fp = calloc(dp->dsize,1);
		fp->disc = dp;
		nv_disc(np, fp, 0);
	}

	/* open GLOBIGNORE node */
	if(np=nv_open("GLOBIGNORE",sh.var_tree,0))
	{
		const Namdisc_t *dp = &SH_GLOBIGNORE_disc;
		Namfun_t *fp = calloc(dp->dsize,1);
		fp->disc = dp;
		nv_disc(np, fp, 0);
	}

	/* set startup files */
	n=0;
	if(!sh_isoption(SH_NOPROFILE))
	{
		if(!sh_isoption(SH_POSIX))
		{
			login_files[n++] = "$HOME/.bash_profile";
			login_files[n++] = "$HOME/.bash_login";
		}
		login_files[n++] = (char*)e_profile;
	}
	sh.login_files = login_files;
	if(sh_isoption(SH_NORC))
		sh.rcfile=0;
	else if(!sh.rcfile)
		sh.rcfile = strdup(sh_mactry("$HOME/.bashrc"));
reinit:
	xtrace = sh_isoption(SH_XTRACE);
	sh_offoption(SH_XTRACE);
	verbose = sh_isoption(SH_VERBOSE);
	sh_offoption(SH_VERBOSE);
	if(np = nv_open("SHELLOPTS", sh.var_tree, NV_NOADD))
		nv_offattr(np,NV_RDONLY);
	iop = sfopen(NULL, bash_pre_rc, "s");
	sh_eval(iop,0);
	if(xtrace)
		sh_offoption(SH_XTRACE);
	if(verbose)
		sh_offoption(SH_VERBOSE);
}
